tests/test_logging.py

```diff
@@ -0,0 +1,321 @@
+import re
+from unittest.mock import patch
+
+import pytest
+import structlog
+
+from cohortextractor import StudyDefinition, codelist, patients
+from cohortextractor.cohortextractor import generate_cohort, generate_measures
+from tests.test_tpp_backend import (  # noqa: F401; We need to import these for setup for the tests that require a tpp database;
+    set_database_url,
+    setup_function,
+    setup_module,
+    teardown_module,
+)
+
+
+@pytest.fixture(name="logger")
+def fixture_logger():
+    """Modify `capture_logs` to keep reference to `processors` list intact,
+    in order to also modify bound loggers which get the list assigned by reference.
+
+    See https://github.com/hynek/structlog/issues/408
+    """
+    cap = structlog.testing.LogCapture()
+    # Modify `_Configuration.default_processors` set via `configure` but always keep
+    # the list instance intact to not break references held by bound loggers.
+    processors = structlog.get_config()["processors"]
+    old_processors = processors[:]  # copy original processors
+    try:
+        processors.clear()  # clear processors list
+        processors.append(cap)  # append the LogCapture processor for testing
+        structlog.configure(processors=processors)
+        yield cap
+    finally:
+        processors.clear()  # remove LogCapture
+        processors.extend(old_processors)  # restore original processors
+        structlog.configure(processors=processors)
+
+
+def get_stats_logs(log_output):
+    return [
```

Comment:
:+1: 
The isinstance() bit is there because in some places either the dataframe or the dataframe index is the log event (I'm pretty sure it's not in cohortextractor code, and it's probably coming from pandas itself, but the LogCapture gets all logs.  Pandas complains about the comparison on a df

Reply:
I'll add a comment